sip/sip-002-smart-contract-language.md

```diff
@@ -0,0 +1,309 @@
+```
+
+Many functions within the language will output values larger than the
+function's input, _however_, these outputs will be bound by
+statically inferable constants. For example, the data function
+_fetch-entry_ will always return an object whose size is equal
+to the specified value type of the map.
+
+A complete proof for the static runtime analysis of smart contracts
+will be included with the implementation of the language.
+
+# Deploying the Smart Contract
+
+Smart contracts on the Stacks blockchain will be deployed directly as
+source code. The goal of the smart contracting language is that the
+code of the contract defines the _ground truth_ about the intended
+functionality of the contract. While seemingly banal, many systems
+chose instead to use a a compiler to translate from a a friendly
+high-level language to a lower-level language deployed on the
+blockchain. Such an architecture is needlessly dangerous. A bug in
+such a compiler could lead to a bug in a deployed smart contract when
+no such bug exists in the original source. This is problematic for
+recovery --- a hard fork to "undo" any should-have-been invalid
+transactions would be contentious and potentially create a rift in the
+community, especially as it will not be easy to deduce which contracts
+exactly were affected and for how long. In contrast, bugs in the VM
+itself present a more clear case for a hard fork: the smart contract
+was defined correctly, as everyone can see directly on the chain, but
+illegal transactions were incorrectly marked as valid.
+
+# Example: Simple Naming System
+
+To demonstrate the expressiveness of this smart contracting language,
+let's look at an example smart contract which implements a simple
+naming system with just two kinds of transactions: _preorder_
+and _register_. The requirements of the system are as follows:
+
+1. Names may only be owned by one principal
+2. A register is only allowed if there is a corresponding preorder
+   with a matching hash
+3. A register transaction must be signed by the same principal who
+   paid for the preorder
+4. A preorder must have paid at least the price of the name. Names
+   are represented as integers, and any name less than 100000 costs
+   1000 microstacks, while all other names cost 100 microstacks.
+
+In this simple scheme, names are represented by integers, but in
+practice, a buffer would probably be used.
+
+```scheme
+(define burn-address '1111111111111111111114oLvT2)
+(define (price-function name)
+  (if (< name 1e5) 1000 100))
+
+(defmap name-map 
+  ((name integer)) ((buyer principal)))
+(defmap preorder-map
+  ((name-hash (buffer 160)) (buyer principal))
+  ((paid integer)))
+
+(define (tx-preorder 
+           (buyer-principal principal)
+           (name-hash (buffer 20))
+           (name-price integer))
+  (if (stacks-transfer!
+        buyer-principal name-price burn-address name-price)
+      (insert-entry! preorder-map
+        (tuple #name-hash name-hash
```

Comment:
Is this a typo? You're passing 4 arguments to `tuple`

Reply:
Nope -- the `#` arguments are just named parameters

Reply:
I think we can revisit the exact syntax we want to use here later, so I'll mark this as resolved for now.